config/kernel.m4

```diff
@@ -942,8 +942,16 @@ dnl # like ZFS_LINUX_TRY_COMPILE, except the contents conftest.h are
 dnl # provided via the fifth parameter
 dnl #
 AC_DEFUN([ZFS_LINUX_TRY_COMPILE_HEADER], [
-	ZFS_LINUX_COMPILE_IFELSE(
-	    [ZFS_LINUX_TEST_PROGRAM([[$1]], [[$2]], [[ZFS_META_LICENSE]])],
-	    [test -f build/conftest/conftest.ko],
-	    [$3], [$4], [$5])
+	AS_IF([test "x$enable_linux_builtin" = "xyes"], [
+		ZFS_LINUX_COMPILE_IFELSE(
+		    [ZFS_LINUX_TEST_PROGRAM([[$1]], [[$2]], [[ZFS_META_LICENSE]])],
+		    [test -f build/conftest/conftest.o],
```

Comment:
Could we can nest AS_IF() to only duplicate the test command rather than duplicating everything here? Other than that, this seems fine to me.

Reply:
I'm not really familiar with m4... like this?

```
ZFS_LINUX_COMPILE_IFELSE(
	    [ZFS_LINUX_TEST_PROGRAM([[$1]], [[$2]], [[ZFS_META_LICENSE]])],
	    AS_IF([test "x$enable_linux_builtin" = "xyes"],
		[test -f build/conftest/conftest.o],
		[test -f build/conftest/conftest.ko]),
	    [$3], [$4], [$5])
```

Reply:
Yes. Also, now that I look at the file, it appears that `ZFS_LINUX_TRY_COMPILE` needs this treatment too.

Reply:
Done!

Reply:
It generates very weird and invalid sh code this way:
```
configure:21940: if test xno = xyes
then :
		test -f build/conftest/conftest.o
else as_fn_nop
		test -f build/conftest/conftest.ko
fi
./configure: eval: line 21942: syntax error: unexpected end of file
```
Probably `ZFS_LINUX_TRY_COMPILE` has its look for reason.

Reply:
If moving AS_IF inside ZFS_LINUX_COMPILE_IFELSE is causing autoools to generate invalid code, then leaving it outside is fine.

As for ZFS_LINUX_TRY_COMPILE, it probably should be patched, or did you find a reason why it should not be patched?

Reply:
They both don't work for the same reason: that generated if-else which tries to test conftest object file is broken, and I didn't find any way to make it work, apart from leaving as it is. I hate autotools TBH :D

Reply:
In that case, is this really all that we need to make sure that the auto tools checks work properly in the kernel builtin case? I am surprised that the other case is not also a problem for that.

Reply:
But `ZFS_LINUX_TRY_COMPILE()` has those checks already -- they did it the same way as it's currently done in this PR -- wrapped the whole macro with if-else, so it always works. I had some non-working macros yesterday and ZFS doesn't build at all if `ZFS_LINUX_TRY_COMPILE()` doesn't work properly, so we're safe.

Reply:
For some reason, my eyes missed that. In that case, this is fine.